src/containerapp/azext_containerapp/_cloud_build_utils.py

```diff
@@ -0,0 +1,226 @@
+        files = [("file", ("build_data.tar.gz", open(tar_file_path, "rb"), "application/x-tar"))]
+        response_file_upload = requests.post(
+            upload_endpoint,
+            files=files,
+            headers=headers)
+        if not response_file_upload.ok:
+            raise ValidationError(f"Error when uploading the file, request exited with {response_file_upload.status_code}")
+        done_spinner = True
+        thread.join()
+
+        # Wait for provisioning state to succeed
+        done_spinner = False
+        thread = display_spinner(f"Waiting for the Cloud Build agent to report status")
+        build_provisioning = True
+        while build_provisioning:
+            build_json_content = BuildClient.get(cmd, builder_name, build_name, resource_group_name)
+            if build_json_content["properties"]["provisioningState"] == "Succeeded":
+                build_provisioning = False
+            time.sleep(2)
+        done_spinner = True
+        thread.join()
+
+        # Initializing logs stream
+        done_spinner = False
+        thread = display_spinner(f"Streaming Cloud Build logs")
+        headers = {'Authorization': 'Bearer ' + token}
+        response_log_streaming = requests.get(
+            log_streaming_endpoint,
+            headers=headers,
+            stream=True)
+        if not response_log_streaming.ok:
+            raise ValidationError(f"Error when streaming the logs, request exited with {response_log_streaming.status_code}")
+        done_spinner = True
+        thread.join()
+
+        # Initializing Buildpack and Stream the logs
+        done_spinner = False
+        thread = display_spinner(f"Buildpack: Initializing")
+        log_execution_phase_pattern = r"===== (.*) =====$"
+        current_phase_logs = ""
+        for line in response_log_streaming.iter_lines():
+            log_line = remove_ansi_characters(line.decode("utf-8"))
+            current_phase_logs += f"{log_line}\n{substatus_indentation}"
+            if "ERROR:" in log_line or "Kubernetes error happened" in log_line:
+                raise CloudBuildError(current_phase_logs)
+
+            log_execution_phase_match = re.search(log_execution_phase_pattern, log_line)
+            if log_execution_phase_match:
+                log_execution_phase = log_execution_phase_match.group(1)
+                # Stop the previous execution phase
+                done_spinner = True
+                thread.join()
+
+                # Start the current one
+                current_phase_logs = ""
+                done_spinner = False
+                thread = display_spinner(f"Buildpack: {log_execution_phase}")
+            log_in_file(log_line, logs_file, no_print=True)
+        done_spinner = True
+        thread.join()
+
+        # TODO: Delete local tar.gz file
+        # os.unlink(tar_file_path)
+
+        final_image = build_json_content["properties"]["destinationContainerRegistry"]["image"]
```

Comment:
will ["properties"]["destinationContainerRegistry"]["image"] also exist in build json? do we need null check?

Reply:
Yes, this value will always be available (and if the API didn't return it, it would be a bug worth getting an exception here), even before the build completes as we can get all these values beforehand.